### ISSUE TITLE
cisco_ios - device hostname > 20 characters causes ReadTimeout exception

### ISSUE THREAD

**Issue by ghmj2417** [ktbyers/netmiko#2718]
Using netmiko 4.0.0 (latest install from pip).  Did not experience this issue using 3.4.0.

Using type `cisco_ios`.  When I have a device with a hostname > 20 characters and try to run `send_config_set`.  I consistently receive an exception ReadTimeout
```
Exception has occurred: ReadTimeout


Pattern not detected: 'abc\\-defgh\\-52\\-52\\-52\\-52.*' in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.
```
When I change the hostname to be 20 characters or less then the commands will be sent successfully. I also tested `send_command_timing` which does not produce an exception.

I did as suggested and increased the read_timeout on the send_config_set but that didn't fix it.  Any ideas or workarounds that I can use (besides changing the device hostnames)?

**Comment by sic162** [comment]
I am having the same issue with netmiko 4.0.0.

Using ```send_config_set``` on a Cisco IOS device and hostname is > 20 characters raises exception:

```
netmiko.exceptions.ReadTimeout:

Pattern not detected: ‘abcdef\\-ghi\\-jkl123\\-mn456\\-7.*’ in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.

You can also look at the Netmiko session_log or debug log for more information.
```

When in config mode on Cisco devices with longer hostnames, a portion of the hostname is cut off at the prompt. for example:

```
Enter configuration commands, one per line.  End with CNTL/Z.
abcdef-ghi-jkl123-mn(config)#
```

I found another post that mentioned using ```set_base_prompt()``` a possible solution after entering config mode with ```config_mode()```... but also looking for suggestions on how this can be resolved when using ```send_config_set```

**Comment by ktbyers** [comment]
Can either of you post the full exception stack trace (so I can better see where in the Netmiko code the exception is occurring).

**Comment by ghmj2417** [comment]
@ktbyers Here you go
```
Traceback (most recent call last):
  ***Removed unnecessary tracing
  File "/mnt/d/scripts/netmiko.py", line 21, in <module>
    router.ssh.send_config_set(config_commands=config_commands)
  File "/home/user/venv3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 2137, in send_config_set
    output += self.config_mode()
  File "/home/user/venv3.8/lib/python3.8/site-packages/netmiko/cisco_base_connection.py", line 47, in config_mode
    return super().config_mode(
  File "/home/user/venv3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 1976, in config_mode
    output += self.read_until_prompt(read_entire_line=True)
  File "/home/user/venv3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 734, in read_until_prompt
    return self.read_until_pattern(
  File "/home/user/venv3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 651, in read_until_pattern
    raise ReadTimeout(msg)
netmiko.exceptions.ReadTimeout: 

Pattern not detected: 'abc\\-defgabcd\\-52\\-52\\-52\\-52.*' in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.

You can also look at the Netmiko session_log or debug log for more information.
```

**Comment by ktbyers** [comment]
Should be fixed here:

https://github.com/ktbyers/netmiko/pull/2720

**Comment by kielmoreland** [comment]
Good afternoon,

Concerning this particular issue, has it been fixed?  While I see the above post that this should have been fixed [#2720](https://github.com/ktbyers/netmiko/pull/2720), the reason I ask is that I am getting a very similar Netmiko error when I am attempting to execute a Nornir runbook:

```
from nornir import InitNornir
from nornir_utils.plugins.functions import print_result
from nornir_napalm.plugins.tasks import napalm_get

nr = InitNornir (
    config_file="config.yml", dry_run=True
)

results = nr.run (
    task=napalm_get, getters=["facts"]
)

print_result(results)
```
When I execute this runbook against my physical infrastructure, I receive the following error:

```
(nornir) kielmoreland@kielmoreland-Aspire-R5-571T:~/Desktop/nornir$ python nornir_get_facts.py
napalm_get**********************************************************************
* Moreland-2960X-Core-sw1 ** changed : False ***********************************
vvvv napalm_get ** changed : False vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv ERROR
Traceback (most recent call last):
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/task.py", line 99, in start
    r = self.task(self, **self.params)
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir_napalm/plugins/tasks/napalm_get.py", line 32, in napalm_get
    device = task.host.get_connection(CONNECTION_NAME, task.nornir.config)
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/inventory.py", line 494, in get_connection
    self.open_connection(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/inventory.py", line 546, in open_connection
    conn_obj.open(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir_napalm/plugins/connections/__init__.py", line 57, in open
    connection.open()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/napalm/ios/ios.py", line 172, in open
    self.device = self._netmiko_open(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/napalm/base/base.py", line 129, in _netmiko_open
    self._netmiko_device.enable()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/cisco_base_connection.py", line 26, in enable
    return super().enable(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 2008, in enable
    output += self.read_until_prompt()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 812, in read_until_prompt
    return self.read_until_pattern(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 721, in read_until_pattern
    raise ReadTimeout(msg)
netmiko.exceptions.ReadTimeout: 

Pattern not detected: 'Moreland\\-2960X\\-C' in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.

You can also look at the Netmiko session_log or debug log for more information.



^^^^ END napalm_get ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* Moreland-LVR-sw1 ** changed : False ******************************************
vvvv napalm_get ** changed : False vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv ERROR
Traceback (most recent call last):
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/task.py", line 99, in start
    r = self.task(self, **self.params)
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir_napalm/plugins/tasks/napalm_get.py", line 32, in napalm_get
    device = task.host.get_connection(CONNECTION_NAME, task.nornir.config)
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/inventory.py", line 494, in get_connection
    self.open_connection(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir/core/inventory.py", line 546, in open_connection
    conn_obj.open(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/nornir_napalm/plugins/connections/__init__.py", line 57, in open
    connection.open()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/napalm/ios/ios.py", line 172, in open
    self.device = self._netmiko_open(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/napalm/base/base.py", line 129, in _netmiko_open
    self._netmiko_device.enable()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/cisco_base_connection.py", line 26, in enable
    return super().enable(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 2008, in enable
    output += self.read_until_prompt()
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 812, in read_until_prompt
    return self.read_until_pattern(
  File "/home/kielmoreland/Desktop/nornir/lib/python3.10/site-packages/netmiko/base_connection.py", line 721, in read_until_pattern
    raise ReadTimeout(msg)
netmiko.exceptions.ReadTimeout: 

Pattern not detected: 'Moreland\\-LVR\\-sw1' in output.

Things you might try to fix this:
1. Adjust the regex pattern to better identify the terminating string. Note, in
many situations the pattern is automatically based on the network device's prompt.
2. Increase the read_timeout to a larger value.

You can also look at the Netmiko session_log or debug log for more information.



^^^^ END napalm_get ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(nornir) kielmoreland@kielmoreland-Aspire-R5-571T:~/Desktop/nornir$ ^C
```
I have seen some individuals suggesting that giving "false" to `fast_cli` in the `groups.yml` file and I have done so:

```
# Nornir groups file
---
moreland_network:
  platform: ios
  connection_options:
    napalm:
      extras:
        optional_args:
          fast_cli: False
```
Therefore, I am at a bit of a loss, and any assistance that may be rendered would be appreciated.
